File: 0_vectors/vector.js
/*
Heru Handika
25 October 2020
Typescript implementation for
vector operation
*/
function add_two_vectors(vec) {
    var result = [0];
    for (var i = 0; i < vec.x.length; i++) {
        result[i] = vec.x[i] + vec.y[i];
    }
    return result;
}
function substract_vectors(vec) {
    var result = [0];
    for (var i = 0; i < vec.x.length; i++) {
        result[i] = vec.x[i] - vec.y[i];
    }
    return result;
}
// Better for small vector size compare to reduce
// method.
function sum_vector(vec) {
    var sum = 0.0;
    for (var i = 0; i < vec.length; i++) {
        sum += vec[i];
    }
    return sum;
}
// Better for large vector size.
function sum_reduce(vec) {
    var sum = vec.reduce(function (accumulator, current) {
        return accumulator + current;
    });
    return sum;
}
function mean_vectors(vec) {
    var mean = sum_vector(vec) / vec.length;
    return mean;
}
function sum_using_of(vec) {
    var sum = 0.0;
    for (var _i = 0, vec_1 = vec; _i < vec_1.length; _i++) {
        var num = vec_1[_i];
        sum += num;
    }
    return sum;
}
var first_vector = { x: [10, 20, 30, 40, 50], y: [1, 2, 3, 4, 5] };
console.log("Vector addition: " +
    add_two_vectors(first_vector));
console.log("Vector substraction: " +
    substract_vectors(first_vector));
console.log("Sum vector x: " +
    sum_vector(first_vector.x));
console.log("Sum using reduce: " +
    sum_reduce(first_vector.x));
console.log("Mean vector x: " +
    mean_vectors(first_vector.x));
console.log("Sum vector using 'of' method: " +
    sum_using_of(first_vector.x));

File: 0_vectors/vector.py
"""
Heru handika
9 October 2020
Vector operation using python
standard library.
"""
from typing import List

def main():
    x: List[int] = [1, 2, 3, 4, 5]
    y: List[int] = [10, 20, 30, 40, 50]

    add_res: List[int] = add_vectors(x, y)
    print(f"Vector addtion: {add_res}")

    sub_res: List[int] = substract_vector(x,y)
    print(f"Vector substraction: {sub_res}")

    mul_res: List[int] = multiply_vector(x, y)
    print(f"Vector multiplication: {mul_res}")

    scal_res: List[int] = multiply_by_scalar(x, 2)
    print(f"Scalar multiplication: {scal_res}")

    sum_x: int = sum_vector(x)
    print(f"Sum vector x: {sum_x}")

    mean_x: float = mean_vectors(x)
    print(f"Mean vector x: {mean_x}")

    dot: int = calculate_dot_product(x, y)
    print(f"Dot product: {dot}")

    sum_of_sq_x: float = calculate_sum_of_square(x)
    print(f"Sum of square x: {sum_of_sq_x}")

def add_vectors(x: List[int], y: List[int]) -> List[int]:
    return [i + j for i, j in zip(x, y) ]

def substract_vector(x: List[int], y: List[int]) -> List[int]:
    return [i - j for i, j in zip(x, y) ]

def multiply_vector(x: List[int], y: List[int]) -> List[int]:
    return [i * j for i, j in zip(x, y) ]

def multiply_by_scalar(vec: List[int], scalar: int) -> List[int]:
    return [scalar * i for i in vec]

# Not a pythonic way. Maybe slow for a large dataset.
def sum_vector(vec: List[int]) -> int:
    sum: float = 0.0;
    for i in vec:
        sum += i
    return sum

# Using python sum function. 
def mean_vectors(vec: List[int]) -> float:
    return float(sum(vec))/len(vec)

def calculate_dot_product(x: List[int], y: List[int]) -> int:
    return sum(multiply_vector(x, y))
 
def calculate_sum_of_square(vec: List[int]) -> float:
    return float(calculate_dot_product(vec, vec))

if __name__ == "__main__":
    main()
